core/dnsserver/address.go

```diff
@@ -94,3 +101,43 @@ const (
 	TransportTLS  = "tls"
 	TransportGRPC = "grpc"
 )
+
+// Build a Validator that rise error if the bound addresses for listeners are overlapping
+// we consider that an unbound address is overlapping all bound addresses for same zome, zame port
+type zoneAddrOverlapValidator struct {
+	registeredAddr map[zoneAddr]zoneAddr // each zoneAddr is registered once by its key (bound and unbound)
+	unboundOverlap map[zoneAddr]zoneAddr // the unbound equiv ZoneAdddr is registered by its original key (bound or unbound)
+}
+
+func newZoneAddrOverlapValidator() *zoneAddrOverlapValidator {
+	return &zoneAddrOverlapValidator{registeredAddr: make(map[zoneAddr]zoneAddr), unboundOverlap: make(map[zoneAddr]zoneAddr)}
+}
+
+func (c *zoneAddrOverlapValidator) registerAndCheck(k *zoneAddr) (same bool, overlap bool, overlapKey *zoneAddr) {
+
+	// we consider there is an overlap if:
+	//  - the current zoneAddr is already registered is already registered for the same zone/port
+	//  - the current zoneAddr is unbound and a non unbound is already registered for the same zone/port
+	//  - the current zoneAddr is bound and an unbound is already registered for the same zone/port
+	if _, ok := c.registeredAddr[*k]; ok {
+		// exact same zone already registered
+		return true, false, nil
+	}
+	uk := zoneAddr{Zone: k.Zone, Address: "", Port: k.Port, Transport: k.Transport}
+	if already, ok := c.unboundOverlap[uk]; ok {
+		if k.Address == "" {
+			// there is already a bound registered
+			return false, true, &already
+		}
+		if _, ok := c.registeredAddr[uk]; ok {
+			// the overlapping zone+port for the unbound addr is already registered
+			return false, true, &uk
+		}
+	} else {
+		// anyway add the unbound equiv of current zoneAddr for future tests
+		c.unboundOverlap[uk] = *k
+	}
+	// anyway add the current zoneAddr for future tests
+	c.registeredAddr[*k] = *k
```

Comment:
when is this code exectuted?

Reply:
when there is no overlap. I need to keep that code.